scripts/babel-plugin-jest-native-promise.js

```diff
@@ -0,0 +1,29 @@
+/**
+ * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+'use strict';
+
+// This plugin exists to make sure that we use a `Promise` that has not been messed with by user code.
+// Might consider extending this to other globals as well in the future
+
+const jestPromise = "(global['jest-promise-stay-away'] || global.Promise)";
+
+module.exports = () => ({
```

Comment:
Since this is only running on Jest's own monorepo packages, I'd personally consider doing this more with something like:

```
module.exports = ({ template }) => {
  const promiseDeclaration = template(`
    var Promise = global['jest-promise-stay-away'] || global.Promise;
  `);

  return {
    name: 'jest-native-promise',
    visitor: {
      ReferencedIdentifier(path, state) {
        if (path.node.name === 'Promise' && !state.injectedPromise) {
          state.injectedPromise = true;
          path.findParent(p => p.isProgram()).unshiftContainer('body', promiseDeclaration());
        }
      },
    },
  };
};
```
This avoids hard-coding the `MemberExpression` and `NewExpression` cases and should work well as long as you're not defining your own `Promise` variable in top-level module scope anywhere, which seems like a reasonable invariant to enforce with or without Babel really.

Reply:
Ah, nice! Way better. Then we should be able to keep using the `Symbol` as well, as it's added at the top level.